# 01. Importing libraries 

In [103]:
# import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Dataframe

In [105]:
# create path
path = r'C:\Users\baidy\Downloads\11-2024 Instacart Basket Analysis'

In [106]:
path

'C:\\Users\\baidy\\Downloads\\11-2024 Instacart Basket Analysis'

In [107]:
#Import dataframe orders
#df_ords= pd.read_csv(r'C:\Users\baidy\Downloads\11-2024 Instacart Basket Analysis\02 Data\Original Data\orders.csv', index_col=False)

In [108]:
df_ords = pd.read_csv(r'C:\Users\baidy\Downloads\11-2024 Instacart Basket Analysis\02 Data\Original Data\orders.csv', index_col=False)

In [109]:
print(df_ords)

         order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                  

# Wrangling Data

In [111]:
# 2.Dropping eval_set column from orders.csv
df_ords=df_ords.drop(columns=['eval_set'])

In [112]:
# changing data type for oeder and user ID in order data frame
df_ords['order_id']=df_ords['order_id'].astype('str')
df_ords['user_id']=df_ords['user_id'].astype('str')

In [113]:
df_ords.value_counts()

order_id  user_id  order_number  order_dow  order_hour_of_day  days_since_prior_order
1         112108   4             4          10                 9.0                       1
3052210   154825   8             1          15                 30.0                      1
3052212   54757    7             2          15                 23.0                      1
3052214   79983    3             0          10                 7.0                       1
3052215   137237   44            1          9                  6.0                       1
                                                                                        ..
2026237   41574    3             0          7                  4.0                       1
2026238   20930    40            3          6                  7.0                       1
2026239   165022   3             2          15                 26.0                      1
202624    5642     12            5          10                 7.0                       1
9999

In [114]:
# 4.the busiest hour for placing order
df_ords['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

# Observation:
the busiest time of the day to place orders is 10.00 am.

In [116]:
# 3.changing the name of the variable 
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace= True)

In [117]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# changing a variables data type

In [119]:
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [120]:
df_ords['order_id']=df_ords['order_id'].astype('str')

In [121]:
df_ords['order_id'].dtype 

dtype('O')

# Import dataframe Products

In [123]:
#Import Dtaframe products
#df_prods= pd.read_csv(r'C:\Users\baidy\Downloads\11-2024 Instacart Basket Analysis\02 Data\Original Data\products.csv',index_col=False)

In [124]:
df_prods= pd.read_csv(r'C:\Users\baidy\Downloads\11-2024 Instacart Basket Analysis\02 Data\Original Data\Products.csv', index_col=False)

# Import dataframe Department

In [126]:
#Import departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data','Original Data','departments.csv'), index_col=False)

# Transposing df_dep

In [128]:
df_dep_t=df_dep.T

# Change header of departments df

In [130]:
# take 1st row of df_dep_t for the header
new_header=df_dep_t.iloc[0]

In [131]:
# copying df_dep_t from row 1 onword
df_dep_t_new=df_dep_t[1:]

In [132]:
#set the header row as the df header
df_dep_t_new.columns=new_header

# Result of the transformation

In [134]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [135]:
df_dep_t_new.shape

(21, 1)

In [136]:
#Assign a name to the index
df_dep_t_new.index.name='department_id'

In [137]:
# reset the index to convert 'department_id' from index to column
df_dep_t_new.reset_index(inplace=True)

In [138]:
df_dep_t_new

department_id department_id       department
0                         1           frozen
1                         2            other
2                         3           bakery
3                         4          produce
4                         5          alcohol
5                         6    international
6                         7        beverages
7                         8             pets
8                         9  dry goods pasta
9                        10             bulk
10                       11    personal care
11                       12     meat seafood
12                       13           pantry
13                       14        breakfast
14                       15     canned goods
15                       16       dairy eggs
16                       17        household
17                       18           babies
18                       19           snacks
19                       20             deli
20                       21          missing

# Observation:
The dataframe is no longer wide and the header has been fixed succesfully

# 4. Creating Data Dictionary

In [141]:
data_dict=df_dep_t_new.to_dict('index')

In [142]:
data_dict

{0: {'department_id': '1', 'department': 'frozen'},
 1: {'department_id': '2', 'department': 'other'},
 2: {'department_id': '3', 'department': 'bakery'},
 3: {'department_id': '4', 'department': 'produce'},
 4: {'department_id': '5', 'department': 'alcohol'},
 5: {'department_id': '6', 'department': 'international'},
 6: {'department_id': '7', 'department': 'beverages'},
 7: {'department_id': '8', 'department': 'pets'},
 8: {'department_id': '9', 'department': 'dry goods pasta'},
 9: {'department_id': '10', 'department': 'bulk'},
 10: {'department_id': '11', 'department': 'personal care'},
 11: {'department_id': '12', 'department': 'meat seafood'},
 12: {'department_id': '13', 'department': 'pantry'},
 13: {'department_id': '14', 'department': 'breakfast'},
 14: {'department_id': '15', 'department': 'canned goods'},
 15: {'department_id': '16', 'department': 'dairy eggs'},
 16: {'department_id': '17', 'department': 'household'},
 17: {'department_id': '18', 'department': 'babies'},
 1

# Observation:
The department_id 4 means the "produce" category.

In [144]:
df_prods[df_prods['department_id']==4]

product_id                                      product_name  aisle_id  \
30             31                                White Pearl Onions       123   
42             43                               Organic Clementines       123   
44             45                                 European Cucumber        83   
65             66                         European Style Spring Mix       123   
88             89                    Yogurt Fruit Dip Sliced Apples       123   
...           ...                                               ...       ...   
49582       49578                                Black Garlic Bulbs       123   
49623       49619                                        Opo Squash        83   
49639       49635  Baby Food Blueberry, Parsnip & Buckwheat Stage 2        83   
49661       49657                                 Cabernet Tomatoes        83   
49687       49683                                    Cucumber Kirby        83   

       department_id  prices  
30                 4     7.5  
42                 4    11.5  
44                 4    14.3  
65                 4    11.7  
88                 4    12.6  
...              ...     ...  
49582              4     8.0  
49623              4    12.7  
49639              4    12.5  
49661              4     8.3  
49687              4    13.2  

[1684 rows x 5 columns]

In [145]:
# 6. Subsetting breakfast item sales
df_breakfast=df_prods[df_prods['department_id']==14]

In [146]:
# Result
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [147]:
# 7. Subsetting products that customer might use to throw dinner parties 
df_dinner_parties=df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [148]:
df_dinner_parties.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

# Observation:
The sales team can use the dataframe df_dinner_parties for their analysis.

In [150]:
# 8. Find the number of rows
df_dinner_parties.shape

(7650, 5)

# Observation:
the new dinner party dataframe has 7650 rows.

In [152]:
# 9. Finding  the information about user_id 1
df_user_id1=df_ords.loc[df_ords['user_id']=='1']
df_user_id1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [153]:
# Step 10: stats about the user's behaviour 
df_user_id1.describe().round(2)

order_number  orders_day_of_week  order_hour_of_day  \
count         11.00               11.00              11.00   
mean           6.00                2.64              10.09   
std            3.32                1.29               3.48   
min            1.00                1.00               7.00   
25%            3.50                1.50               7.50   
50%            6.00                3.00               8.00   
75%            8.50                4.00              13.00   
max           11.00                4.00              16.00   

       days_since_prior_order  
count                   10.00  
mean                    19.00  
std                      9.03  
min                      0.00  
25%                     14.25  
50%                     19.50  
75%                     26.25  
max                     30.00

# Observation:
 The user id1 only places order from Sunday to Wednesday. most of the orders placed at 8 am or earlier,that user places orders on average every 19 days, never spend more than 30 days withoutplacing orders.The user placed 2 orders at same day.

# Export Data

In [156]:
# 12. Export df_ords dataframe as "orders_wrangled.csv" in "Prepared Data" folder
df_ords.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'))

In [157]:
# 12. Export departments dataframe
df_dep_t_new.to_csv(os.path.join(path,'02 Data','Prepared Data','departments_wrangled.csv'))